### PR TITLE
S3 relay interface

### PR DESCRIPTION
**Pull request by cody-littley**
## Why are these changes needed?

This PR adds utilities for uploading and downloading files from S3 in a way that supports breaking files into smaller ones.

## Checks

- [x] I've made sure the lint is passing in this PR.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- [ ] I've checked the new test coverage and the coverage percentage didn't drop.
- Testing Strategy
   - [x] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: go.mod
Comment (dmanc) [inline go.mod:9]: Is it possible to do it with just v2 library?

File: relay/dataplane/config.go
Comment (dmanc) [inline relay/dataplane/config.go:24]: Should we set the default as number of cpus in machine? Like by using gomaxprocs

**Review by dmanc — COMMENTED** [review]
Looks great, some initial comments

File: relay/dataplane/s3_client.go
Comment (ian-shim) [inline relay/dataplane/s3_client.go:28]: I think what you're doing here can be simplified & abstracted out by using `workerpool` (ex. https://github.com/Layr-Labs/eigenda/blob/master/api/clients/retrieval_client.go#L181)

**Review by ian-shim — COMMENTED** [review]
Looks good! Few comments

File: relay/dataplane/s3_client.go
Comment (cody-littley) [reply relay/dataplane/s3_client.go:28]: This is a neat little library, good to know about. Done.

File: relay/dataplane/config.go
Comment (cody-littley) [reply relay/dataplane/config.go:24]: I've turned this into two parameters: `ParallelismFactor` and `ParallelismConstant`. The total number of workers is set equal to the formula `ParallelismFactor * numCores + ParallelismConstant`.

We will want to have a lot more workers than cores, since most of the time the workers are blocked on IO tasks. This allows us to set a sane default that uses a good number of workers as the number of cores grows.

File: common/aws/cli.go
Comment (ian-shim) [inline common/aws/cli.go:22]: nit: start comment with the name of the field..

File: common/aws/cli.go
Comment (ian-shim) [inline common/aws/cli.go:41]: Why not just use a single field to explicitly define the number of workers?

File: common/aws/s3/client.go
Comment (ian-shim) [inline common/aws/s3/client.go:208]: This pool is shared across all ongoing uploads and downloads. Which means when there are many uploads/downloads in flight, it can build backpressure. Is that problematic?

**Review by ian-shim — APPROVED** [review]
lgtm! 
Please resolve the test failures before merging

File: common/aws/cli.go
Comment (cody-littley) [reply common/aws/cli.go:22]: fixed

File: common/aws/cli.go
Comment (cody-littley) [reply common/aws/cli.go:41]: This is based on @dmanc's request to scale the number of workers based on the number of cores on the machine. The pattern here allows for the user to specify either a fixed number of threads or a number that varies with the number of cores.

If you think this is over complicated I'm willing to go back to having a constant number of worker threads.

File: common/aws/s3/client.go
Comment (cody-littley) [reply common/aws/s3/client.go:208]: My previous iteration had a config parameter that allowed the client to be have a configurable sized work queue (i.e. when I was using my own implementation of `workerpool`. Unfortunately, the `workerpool` library does not allow us to override the channel used to send data to the workers.

From the implementation:
```
workerQueue: make(chan func())
```

By default, channels have a buffer size of 0. This means that if the number of read/write tasks exceed the number of available workers, the caller will block until all tasks are accepted by a worker. This will provide back pressure if more work is scheduled than there are workers to handle that work.

Are you ok with the way this is configured by default? If not, we will probably need our own implementation of `workerpool`.

**Pull request merged**
